karapace/serialization.py

```diff
@@ -165,59 +170,70 @@ def __init__(
         self.subject_name_strategy = NAME_STRATEGIES[name_strategy]
         self.registry_client: Optional[SchemaRegistryClient] = registry_client
         self.ids_to_schemas: Dict[int, TypedSchema] = {}
-        self.schemas_to_ids: Dict[str, int] = {}
+        self.schemas_to_ids: Dict[str, SchemaId] = {}
 
     async def close(self) -> None:
         if self.registry_client:
             await self.registry_client.close()
             self.registry_client = None
 
-    def get_subject_name(self, topic_name: str, schema: str, subject_type: str, schema_type: SchemaType) -> str:
-        schema_typed = ParsedTypedSchema.parse(schema_type, schema)
-        namespace = "dummy"
-        if schema_type is SchemaType.AVRO:
-            if isinstance(schema_typed.schema, avro.schema.NamedSchema):
-                namespace = schema_typed.schema.namespace
-        if schema_type is SchemaType.JSONSCHEMA:
-            namespace = schema_typed.to_dict().get("namespace", "dummy")
-        #  Protobuf does not use namespaces in terms of AVRO
-        if schema_type is SchemaType.PROTOBUF:
-            namespace = ""
-        return f"{self.subject_name_strategy(topic_name, namespace)}-{subject_type}"
+    def get_subject_name(
+        self, topic_name: str, schema: Union[ParsedTypedSchema, None], subject_type: str, schema_type: SchemaType
+    ) -> Subject:
+        namespace = None
+
+        if schema is not None:
+            namespace = "dummy"
+            if schema_type is SchemaType.AVRO:
+                if isinstance(schema.schema, avro.schema.NamedSchema):
+                    namespace = schema.schema.namespace
+            if schema_type is SchemaType.JSONSCHEMA:
+                namespace = schema.to_dict().get("namespace", "dummy")
+            #  Protobuf does not use namespaces in terms of AVRO
+            if schema_type is SchemaType.PROTOBUF:
+                namespace = ""
+        return Subject(f"{self.subject_name_strategy(topic_name, namespace)}-{subject_type}")
```

Comment:
issue: It looks like `namespace` can now be `None` when we reach the `return`, but it doesn't look like the (implicit) strategy interface supports `record_name: None`. So this doesn't look right to me.

Reply:
It's ugly but as the software is now this is working [issue here](https://github.com/aiven/karapace/issues/667).
We are always only using the `topic_name` strategy, isn't possible for the user to use another strategy. So if you look the specific implementation:
```python
NAME_STRATEGIES = {
    "topic_name": topic_name_strategy,
    "record_name": record_name_strategy,
    "topic_record_name": topic_record_name_strategy,
}
```

--->

```python
def topic_name_strategy(topic_name: str, record_name: str) -> str:  # pylint: disable=unused-argument
    return topic_name
```

you know why it works 😞